Paper: Composable Multi-Threading for Python Libraries

antalmal opened the pull request:
Initial skeleton for the full paper submisstion

antalmal commented:
I'll substentially modify the paper till the end of this week

Carreau commented:
Thanks for the head's up. Haven't had a chance to look at it yet.

antalmal commented:
Updated.. but still far from completion :-\

sbenthall commented:
The build server is not building this PDF. Does the make_paper.sh script work for you locally? cc @sarostru

antalmal commented:
I didn't try to build it yet. I'll look how to fix it

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+Over-subscription
+-----------------
+Multi-processing parallelism in Python might be inefficient due to memory-related overhead. On the other hand, multi-threaded parallelism is know to be more efficient but with Python, it suffers from the GIL. However, when it comes to numeric computations, most of the time is spent in native codes where the GIL can easily be released. This is why libraries such as Dask and Numba can use multi-thr ...
```
Avoid modal verbs that express possibility ("might be") and stick to what can be substantiated. You could reword to `Multi-processing parallelism in Python are prone to inefficiency due to memory-related overhead`.

Regarding suffering from the GIL, you might want to be more specific - that the GIL nullifies most of the benefits of threading.

"But being used together in a nested way..." is very awkward.

Consider italicising function names for clarity.

The last sentence here is excellent, but could probably be reworked for more emphasis, since it's a key point.

kynan replied:
The GIL should at least briefly be explained, preferably with a reference.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+   with each other leading to overhead and inefficiency. We'll show how to fix this problem.
+
+.. class:: keywords
+   Multi-threading, Over-subscription, Parallel Computations, Parallelism, Threading, Dask, Joblib, Numpy, Scipy
```
Add GIL to the keyword list.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. class:: abstract
+
+   Python is popular among numeric communities that value it for easy to use
```
Abstract needs some more meat to it - such as a problem statement. It should contain:
* What is the problem?
* Why does that matter?
* What can we do about it?
* What is innovative about our solution?

dalippa replied:
If your solution is not universal (and requires hardware-specific capabilities), that __must__ go in the abstract. The title hints at a *universal* approach to solving the problem.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+-----------------
+Multi-processing parallelism in Python might be inefficient due to memory-related overhead. On the other hand, multi-threaded parallelism is know to be more efficient but with Python, it suffers from the GIL. However, when it comes to numeric computations, most of the time is spent in native codes where the GIL can easily be released. This is why libraries such as Dask and Numba can use multi-thr ...
+
+Another example are Numpy/Scipy libraries. When they are accelerated using Intel(R) Math Kernel Library (Intel MKL) like the ones shipped as part of Intel(R) Distribution for Python. Intel MKL is usually threaded using OpenMP which is known for not easily co-existing even with itself. In particular, OpenMP threads keep spin-waiting after the work is done -- which is usually necessary to reduce wo ...
```
"OpenMP which is known for not easily co-existing even with itself" and following substantiation requires a source/citation - even if it's source code that you are summarizing. I'm not quite seeing where OpenMP is not coexisting nicely with itself.

"spin-waiting" is not quite accurate. OpenMP uses spin-locks or "busy waiting". This kind of hodge-podges the two.

"But it plays against it with another thread pool..." is very awkward and seems verbose. You might want to break this up some. First, describe how OpenMP works; then, you can contrast it with Numba.

kynan replied:
"When they are accelerated using Intel(R) Math Kernel Library (Intel MKL) like the ones shipped as part of Intel(R) Distribution for Python" is not a proper sentence. I suppose this is a continuation of the previous sentence?

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+Another example are Numpy/Scipy libraries. When they are accelerated using Intel(R) Math Kernel Library (Intel MKL) like the ones shipped as part of Intel(R) Distribution for Python. Intel MKL is usually threaded using OpenMP which is known for not easily co-existing even with itself. In particular, OpenMP threads keep spin-waiting after the work is done -- which is usually necessary to reduce wo ...
+
+Though overheads from linear over-subscription (e.g. 2x) are not always visible on the application level for small systems and can be tolerable in many cases. But the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region. This results in quadratic over-subscription which ruins multi-threaded performance on systems with sign ...
```
First sentence is unclear because it is a sentence fragment.

You might want to describe the situation something like "Linear increases in overhead, such as a factor of 2, are not always visible in small systems, and the cost can be tolerable in many cases. However, if each task executes an OpenMP parallel region, the number of over-subscriptions increases quadratically -- ruining multi-threaded performance on systems with significant numbers of threads." This is especially clear if you describe the linear nature of executing an OpenMP parallel region.

kynan replied:
The assumptions are not entirely clear here: You need to explain that by default OpenMP uses as many threads as logical cores on the system. That is what leads to the quadratic over-subscription.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+Though overheads from linear over-subscription (e.g. 2x) are not always visible on the application level for small systems and can be tolerable in many cases. But the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region. This results in quadratic over-subscription which ruins multi-threaded performance on systems with sign ...
+
+Especially, over-subscription is bad for Many Integrated Core (MIC) systems such as Intel(R) Xeon Phi which support more than 2 hundred hardware threads. In such big systems, sometimes it is not even possible to create as many software threads as the number of hardware threads multiplied by itself. It will just eat up all the available RAM.
```
First sentence is awkward. You probably should add a citation on the hardware specification.

"It will just eat up all the RAM". "It" is not particularly well defined, nor is why the RAM would be consumed. Threads are much lighter weight and don't have the same RAM utilization that forked processes can have.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+Threading Composability
+-----------------------
+Our approach to solve these co-existence problems is to share one thread pool among all the program modules and native libraries so that one task scheduler will take care of composability between them. Intel(R) Threading Building Blocks (Intel TBB) library works as such a task scheduler in our solution. Intel TBB is a wide-spread and recognized C++ library for enabling multi-core parallelism. It  ...
```
You mention that your solution is for Intel TBB. Is this only supported on x86 processors? What about a subset of Intel processors?

You might want to define composability.

kynan replied:
Could another task scheduler be used? What makes TBB more suitable than alternatives?

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+-----------------------
+Our approach to solve these co-existence problems is to share one thread pool among all the program modules and native libraries so that one task scheduler will take care of composability between them. Intel(R) Threading Building Blocks (Intel TBB) library works as such a task scheduler in our solution. Intel TBB is a wide-spread and recognized C++ library for enabling multi-core parallelism. It  ...
+
+In the Intel� Distribution for Python* 2017 Beta, I introduce an experimental module which unlocks opportunities for additional performance for multi-threaded Python programs by enabling threading composability between two or more thread-enabled libraries. Threading composability can accelerate programs by avoiding inefficient threads allocation discussed above.
```
Be consistent with your narrative style (especially 1st person). Since you are the sole author, I think utilizing "I" throughout as well as "we" are reasonable, provided they are used consistently.

kynan replied:
Avoid the first person. Does that * refer to a non-existent footnote?

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+In the Intel� Distribution for Python* 2017 Beta, I introduce an experimental module which unlocks opportunities for additional performance for multi-threaded Python programs by enabling threading composability between two or more thread-enabled libraries. Threading composability can accelerate programs by avoiding inefficient threads allocation discussed above.
+
+The module implements Pool class with the standard Python interface using Intel� TBB which can be used to replace Python�s ThreadPool. Thanks to the monkey-patching technique implemented in class Monkey, no source code change is needed in order to enable single thread pool across different Python modules. It also enables TBB-based threading layer for Intel(R) MKL which automatically enables compo ...
```
When referring to words in your code, please use italics or some other mechanism to improve readability, such as "The module implements the class *Pool* with the..."

dalippa replied:
You need citations for the classes you refer to, as well as citations to figures/code blocks.

kynan replied:
Do all packages you list in the introduction actually use the standard Python ThreadPool? What happens if they do not?

antalmal replied:
Yes, they do. If some packages don't, they are subjects for the same issue of over-subscription.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+Using TBB module
+----------------
+For our first experiment, we need Intel� Distribution for Python to be installed along with Dask library:
```
Grammatical error.

Carreau replied:
You might need to shortly introduce what Dask is. It is still relatively recent so not everyone might be aware of what it does.

dalippa replied:
An APA citation that also links to Dask's source code would also be welcome.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. code-block:: shell
+
+    source <path to Intel� Distribution for Python*>/bin/pythonvars.sh
```
You might also want to provide a hyperlink for downloading.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+:author: Anton Malakhov
```
Build failed; no PDF: http://zibi.bids.berkeley.edu:5000/#

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
If you can use a caption on the code, it might be good to use that here.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
Fix TODOs.

Avoid colloquialism like "That's it!" in a conference paper. Consider a sentence that explains that the library is easy to install and improves performance. Also, you might want to consider phrasing the reduction in compute time with the increase of productivity (50% reduction = 100% increase in productivity). I think it's easier to compare performance improvements the opposite way that you described here.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
Missing diagram

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
Probably should go into a "Future Work" section, stating that enhancing Numba with TBB tasks is promising, and could be used to enhance other such libraries.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
Fix TODO.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
Disclaimers should go into a separate section "Limitations" and cover all such limitations (hardware-specific requirements, blocking I/O, etc.)

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
Needs a conclusion, and is missing citations

dalippa commented:
Paper is intriguing, but is still woefully incomplete.

High-level comments:
* Missing sections (future work, methods, conclusion, etc)
* No citations for substantiation
* Unclear narrative
* Missing diagrams
* Does not compile

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+.. class:: abstract
+
+   Python is popular among numeric communities that value it for easy to use
+   number crunching modules like Numpy/Scipy, Dask, Numba, and many others.
```
One __really__ important point that should be made here is that CPU speeds have been leveling off quite a bit (http://i.imgur.com/NFysh.png) but applications have not been benefiting from the nine-year trend of multi-core processors that are available in commodity hardware. It is also very important to note that the number of cores per processor hasn't increased much in commodity hardware either.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
Presumably TBB's package entry point does the monkey patching described above?

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
Again, avoid 1st person.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
You also need to clearly state the licenses of the mentioned Intel products and their cost. Presumably those are neither open source nor free.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
I've seen many papers with multiple authors write in 1st person plural. I think 1st person singular is reasonable, given there is only one author; however, a 3rd person narrative is most preferred.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,87 @@
+:institution: Intel Corporation
+
+-----------------------------------------------
+COMPOSABLE MULTI-THREADING FOR PYTHON LIBRARIES
```
Please be consistent (with other papers) with regard to title capitalization.

kynan commented:
This paper discusses the (very relevant!) issue of thread oversubscription when using nested regions of parallelism and describes a potential solution using Intel's TBB as a task scheduler by monkey patching Python's standard ThreadPool.

However the paper is a mere skeleton. All results, the conclusion and citations are missing. Furthermore the existing text needs substantial improvement. Finally it is unclear whether the proposed solution requires (costly) commercial licenses from Intel.

I would really like to see this paper finished!

suchow commented:
Here's some advice from Proselint (http://github.com/amperser/Proselint), a Python-based linter for prose, with two false alarms removed:

```
72:2: misc.annotations Annotation left in text.
79:2: misc.annotations Annotation left in text.
83:2: misc.annotations Annotation left in text.
```

antalmal commented on papers/anton_malakhov/00_amalakhov.rst:
@kynan Intel TBB is open source and Intel Distribution for Python along with Intel MKL are free.

antalmal commented on papers/anton_malakhov/00_amalakhov.rst:
Thanks everyone for the detailed review! I didn't have enough time to work on this text yet. But now, I'll be working on the paper full-time during all the next week and even more since I need to say the same things in my talk anyway.

antalmal commented:
Thanks everyone for the detailed review! I apologize that I didn't have enough time to work on this text yet (Intel closes our site in IL, so I'm moving to TX). But now, I'll be working on the paper full-time during all the next week and even more since I need to say the same things in my talk anyway.

Carreau commented on papers/anton_malakhov/00_amalakhov.rst:
Side questions, how does TBB interact with pseudo random number generation ? For many people, it is important to have reproducible sequences if you set a specific seed, so the point of faster performances can become irrelevant if you cannot check exactness of your results.

Carreau commented:
> Thanks everyone for the detailed review! I apologize that I didn't have enough time to work on this text yet (Intel closes our site in IL, so I'm moving to TX). But now, I'll be working on the paper full-time during all the next week and even more since I need to say the same things in my talk anyway.

Good luck for your moving.

sbenthall commented:
We're getting this build error on our server:

https://gist.github.com/sbenthall/4b7a392252105fbd98509f608c142911

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+   Python is popular among numeric communities that value it for easy to use number crunching modules like Numpy/Scipy, Dask, Numba, and many others.
+   These modules often use multi-threading for efficient parallelism (on a node) in order to utilize all the available CPU cores.
+   But being used together in one application, their threads can interfere with each other leading to overheads and inefficiency.
+   The lost performance can still be recovered if all the multi-threaded parties are coordinated.
```
Isn't it rather the case that you don't lose the performance in the first place if you coordinate?

antalmal replied:
it is what I'm trying to say here

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+   These modules often use multi-threading for efficient parallelism (on a node) in order to utilize all the available CPU cores.
+   But being used together in one application, their threads can interfere with each other leading to overheads and inefficiency.
+   The lost performance can still be recovered if all the multi-threaded parties are coordinated.
+   This paper describes usage of Intel |R| Threading Building Blocks (Intel |R| TBB), an open-source cross-platform library for multi-core parallelism, as coordination layer for Python libraries which helps to extract additional performance for numeric applications on multi-core system.
```
Coordination layer for *Python threads*, no? systems is missing an s

antalmal replied:
No, it is wider than coordination of Python threads only because libraries like numpy, scipy, and numba matter a lot as well another C extensions, and it would be hard to say that they use Python to create threads. thanks.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+
+Over-subscription
+-----------------
+Multi-processing parallelism in Python is prone to inefficiency due to memory-related overheads. On the other hand, multi-threaded parallelism is know to be more efficient but with Python, it suffers from the global interpreter lock (GIL) :cite:`gil` which prevents scaling of Python programs. However, when it comes to numeric computations, most of the time is spent in native codes where the GIL c ...
```
2nd sentence sounds stilted, needs reformulating. You're already talking about Python in sentence 1, no need for the "but".

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+-----------------
+Multi-processing parallelism in Python is prone to inefficiency due to memory-related overheads. On the other hand, multi-threaded parallelism is know to be more efficient but with Python, it suffers from the global interpreter lock (GIL) :cite:`gil` which prevents scaling of Python programs. However, when it comes to numeric computations, most of the time is spent in native codes where the GIL c ...
+
+Another example are Numpy/Scipy libraries. When they are accelerated using Intel |R| Math Kernel Library (Intel |R| MKL) like the ones shipped as part of Intel |R| Distribution for Python. Intel MKL is by default is threaded using OpenMP which is known for its inherent restrictions. In particular, OpenMP threads keep busy-waiting after the work is done - which is usually useful to reduce work dis ...
```
Split the sentence with the ; properly in 2. Clarify that OpenMP keeps the threads active so they can be reused in subsequent parallel regions (does not become clear from your text).

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+
+Another example are Numpy/Scipy libraries. When they are accelerated using Intel |R| Math Kernel Library (Intel |R| MKL) like the ones shipped as part of Intel |R| Distribution for Python. Intel MKL is by default is threaded using OpenMP which is known for its inherent restrictions. In particular, OpenMP threads keep busy-waiting after the work is done - which is usually useful to reduce work dis ...
+
+Though overheads from linear over-subscription (e.g. 2x) are not always visible on the application level for small systems and can be tolerable in many cases when the work for parallel region is big enough. But the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region. This results by default in quadratic over-subscription  ...
```
1st sentence (still) a fragment.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+
+Though overheads from linear over-subscription (e.g. 2x) are not always visible on the application level for small systems and can be tolerable in many cases when the work for parallel region is big enough. But the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region. This results by default in quadratic over-subscription  ...
+
+Especially, over-subscription is bad for Many Integrated Core (MIC) systems such as Intel |R| Xeon Phi which support more than 2 hundred hardware threads. In such big systems, sometimes it is not even possible to create as many software threads as the number of hardware threads multiplied by itself. It will just eat up all the available RAM.
```
Still not convinced that RAM is the issue here...

antalmal replied:
on 64-thread system like I also use for testing, `ulimit -s` gives 8M stack size. 64*64*8MB = 32GB. Easy. Though, I agree with you that there are reasons beside memory. But results is the same:
```
OMP: Error #34: System unable to allocate necessary resources for OMP thread:
OMP: System error #11: Resource temporarily unavailable
OMP: Hint: Try decreasing the value of OMP_NUM_THREADS.
```

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+
+Our approach to solve these co-existence problems is to share one thread pool among all the program modules and native libraries so that one user-level task scheduler will take care of composability between them. Intel |R| Threading Building Blocks (Intel TBB) library works as such a task scheduler in our solution, see Figure :ref:`components`. Intel |R| TBB is open-source, cross-platform, recogn ...
+
+In the Intel |R| Distribution for Python* 2017 Beta and later as part of Intel |R| TBB release 4.4 Update 4, I introduce an experimental module which unlocks opportunities for additional performance for multi-threaded Python programs by enabling threading composability between two or more thread-enabled libraries. Threading composability can accelerate programs by avoiding inefficient threads all ...
```
What does the * indicate? Avoid 1st person.

antalmal replied:
Strange Intel trademarks rules..

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+
+Figure :ref:`qrpic` shows times (lower is better) acquired on 32-core (no HT) machine with 64GB RAM. By default, Dask version runs worse than Numpy version because 10 outermost tasks end up calling 10 OpenMP-based parallel regions which creates 10 times more threads than available hardware resourses.
+
+The second command runs this benchmark with innormost OpenMP parallelism disabled. It results in the worst performance for Numpy version since everything is now serialized. And Dask version is not able to close the gap completely since it has only 10 tasks which can run in parallel while Numpy with parallel MKL is able to utilize the whole machine with 32 threads.
```
What performance do you get by actually running an optimal 32 tasks?

antalmal replied:
I'll write about that. When there are enough outermost parallelism to fully load the machine, TBB does not differ from serializing of the nested parallel regions. It can only help by load-balancing the work at the end. But for examples I use, it does not give any visible difference. The whole point of TBB as coordination layer is that it can  add threads to parallel region while OpenMP is unable to do so by its specification.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,124 @@
+   
+Numba
+-----
+Another area where we applied Intel TBB is Numba. I replaced multi-threading runtime used by original Numba with implementation based on TBB tasks. It improved performance even without nested parallelism:
```
Avoid 1st person

kynan commented:
There are still lots of grammatical errors throughout the text. Do you have a native speaker colleague to proofread this?

antalmal commented:
Sure, I'll run it through lots of checks before publishing. Meanwhile, I'm trying to add more meaning and data to the paper.

dalippa commented:
I agree with @kynan . @antalmal please let me know when you're ready for a re-review.

antalmal commented:
Current status: I filled the paper with almost everything I wanted to say. There are few benchmarks which probably will not be included into the paper because of the technical difficulties. Though, on the other hand, I have additional data for QR benchmark but I'm hesitant to include it as well.

Anyway, I need more time for rephrasing first sections, adding references, and running it through spell and grammar checkers. But I hope the decision whether to include it into Scipy proceedings if required now, can be positive basing on the essential&experimental parts.

@kynan @dalippa

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+   Python is popular among numeric communities that value it for easy to use number crunching modules like Numpy/Scipy, Dask, Numba, and many others.
+   These modules often use multi-threading for efficient parallelism (on a node) in order to utilize all the available CPU cores.
+   But being used together in one application, their threads can interfere with each other leading to overheads and inefficiency.
```
avoid starting sentence with conjunction (or can combine into previous sentence)

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+Composable Multi-Threading for Python Libraries
+-----------------------------------------------
+
+.. class:: abstract
```
I think it's very salient that multi-processing is currently sub-optimal in Python. It's part of the problem statement and should be mentioned in the abstract.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Introduction to Over-subscription
+---------------------------------
+Multi-processing parallelism in Python is prone to inefficiency due to memory-related overheads. On the other hand, multi-threaded parallelism is know to be more efficient but with Python, it suffers from the global interpreter lock (GIL) :cite:`gil` which prevents scaling of Python programs. However, when it comes to numeric computations, most of the time is spent in native codes where the GIL c ...
```
"it suffers from the GIL" is not quite accurate. Probably better as "it suffers from the limitations of the global interpreter lock (GIL)".

dalippa replied:
"But when used together..." This is very verbose. Instead consider something more general such as, "however, these libraries do not coordinate the creation or pooling of threads, and leads to the situation where..."

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+---------------------------------
+Multi-processing parallelism in Python is prone to inefficiency due to memory-related overheads. On the other hand, multi-threaded parallelism is know to be more efficient but with Python, it suffers from the global interpreter lock (GIL) :cite:`gil` which prevents scaling of Python programs. However, when it comes to numeric computations, most of the time is spent in native codes where the GIL c ...
+
+Another example are Numpy/Scipy libraries. For example in  Intel |R| Distribution for Python, they are accelerated using Intel |R| Math Kernel Library (Intel |R| MKL). Intel |R| MKL is threaded by default using OpenMP which is known for its inherent restrictions. For instance, OpenMP keeps the threads active so they can be reused in subsequent parallel regions. Usually, this is useful approach to ...
```
in "known for its inherent restrictions" you might want to summarize the restrictions before diving into one or more examples. You claim there are multiple restrictions and are well-known, but without citation (of what they are, that it's a well known problem, etc)

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Another example are Numpy/Scipy libraries. For example in  Intel |R| Distribution for Python, they are accelerated using Intel |R| Math Kernel Library (Intel |R| MKL). Intel |R| MKL is threaded by default using OpenMP which is known for its inherent restrictions. For instance, OpenMP keeps the threads active so they can be reused in subsequent parallel regions. Usually, this is useful approach to ...
+
+Though overheads from linear over-subscription (e.g. 2x) are not always visible on the application level (especially for small systems) and can be tolerated in many cases when the work for parallel region is big enough. But the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region. This results by default in quadratic over- ...
```
overhead should be singular.

Sentence "Though overheads from ..." is a sentence fragment.

"This results by default..." is awkward. Is it default settings that cause the quadratic over-subscription?

Instead of breaking up "In some big systems, sometimes, ..." consider rewording. It's also a run-on sentence.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+   Intel |R| Threading Building Blocks is used as a common coordinating runtime for different Python modules. :label:`components`
+
+Our approach to solve these co-existence problems is to share single thread pool among all the program modules and native libraries so that one user-level task scheduler will take care of composability between them. Intel |R| Threading Building Blocks (Intel |R| TBB) library works as such a task scheduler in our solution, see Figure :ref:`components`. Intel |R| TBB is an open-source, cross-platfo ...
```
Using a single thread pool can make an application prone to starvation. I am not seeing how you are addressing that.

antalmal replied:
I hope I addressed this concern discussing asynchronous vs. blocking I/O in the Limitations section

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Our approach to solve these co-existence problems is to share single thread pool among all the program modules and native libraries so that one user-level task scheduler will take care of composability between them. Intel |R| Threading Building Blocks (Intel |R| TBB) library works as such a task scheduler in our solution, see Figure :ref:`components`. Intel |R| TBB is an open-source, cross-platfo ...
+
+In the Intel |R| Distribution for Python* 2017 Beta and later as part of Intel |R| TBB release 4.4 Update 4, I introduce an experimental module which unlocks opportunities for additional performance for multi-threaded Python programs by enabling threading composability between two or more thread-enabled libraries. Threading composability can accelerate programs by avoiding inefficient threads all ...
```
Your title and paper use "thread composability" often and it seems like (necessary) jargon, but is not defined. It should be defined, at least at a high level, in this section.

dalippa replied:
threads -> thread

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+In the Intel |R| Distribution for Python* 2017 Beta and later as part of Intel |R| TBB release 4.4 Update 4, I introduce an experimental module which unlocks opportunities for additional performance for multi-threaded Python programs by enabling threading composability between two or more thread-enabled libraries. Threading composability can accelerate programs by avoiding inefficient threads all ...
+
+The TBB module implements :code:`Pool` class with the standard Python interface using Intel |R| TBB which can be used to replace Python's *ThreadPool*. Thanks to the monkey-patching technique implemented in class :code:`Monkey`, no source code change is needed in order to enable single thread pool across different Python modules. It also enables TBB-based threading layer for Intel |R| MKL which a ...
```
Not everyone may be familiar with monkey-patching. You might want to briefly describe that technique.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+    # install Dask
+    conda install dask
+
+Now, let's write a simple program using Numpy that validates QR decomposition by multiplying resulted components and comparing result agianst original input:
```
Avoid informal usage such as this. Instead, maybe "Consider a simple program below that uses NumPy to validate..."

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+    assert(test)
+    print(time.time() - t0)
+
+And same program using Dask:
```
sentence fragment

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Here, Dask splits the array into 10 chunks and processes them in parallel using multiple threads. But each Dask task executes the same Numpy matrix operations which are accelerated using Intel |R| MKL under the hood and thus multi-threaded by default. This combination results in nested parallelism, i.e. when one parallel component calls another component which is also threaded.
+
+Let's run it in 3 different modes:
```
Informal usage. Consider "Here is an example of X's 3 different modes"

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+   
+   Execution times for QR validation example. :label:`qrpic`
+
+Figure :ref:`qrpic` shows times (lower is better) acquired on 32-core (no HT) machine with 64GB RAM. By default, Dask version runs worse than Numpy version because 10 outermost tasks end up calling 10 OpenMP-based parallel regions which creates 10 times more threads than available hardware resourses.
```
"lower is better" should be in the caption, not here. However, with respect to time, you may be able to leave this out altogether.

"By default..." is grammatically incorrect and awkward.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Figure :ref:`qrpic` shows times (lower is better) acquired on 32-core (no HT) machine with 64GB RAM. By default, Dask version runs worse than Numpy version because 10 outermost tasks end up calling 10 OpenMP-based parallel regions which creates 10 times more threads than available hardware resourses.
+
+The second command runs this benchmark with innormost OpenMP parallelism disabled. It results in the worst performance for Numpy version since everything is now serialized. And Dask version is not able to close the gap completely since it has only 10 tasks which can run in parallel while Numpy with parallel MKL is able to utilize the whole machine with 32 threads. The reason why only 10 tasks wer ...
```
innormost -> innermost

"And Dask version..." is grammatically incorrect.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+The last command demostrates how Intel TBB can be enabled as orchestrator of multi-threaded modules. TBB module runs the benchmark in context of :code:`with TBB.Monkey():` which replaces standard Python *ThreadPool* class used by Dask and also switches MKL into TBB mode. Numpy with TBB shows more than double time comparing to default Numpy run. This happens because TBB-based threading in MKL is n ...
+
+.. [#] For more complete information about compiler optimizations, see our Optimization Notice
```
you can cite the optimization notice explicitly here and link to it.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+    Case study results: Generation of User Recommendations. :label:`casestudy`
+
+The leftmost result was acquired on pure, non-accelerated Python which comes by default on Fedora 23. It is the base. Running the same application without modifications with Intel |R| Distribution for Python* results in 17x times speedup. One reason for this performance increase is that Intel |R| MKL runs computations in parallel. Thus for sake of experiment, outermost parallelism was implemented ...
```
x15 times faster? Please be consistent with describing metrics. 17x = 17 times (17x times is incorrect).

dalippa commented:
There are still a lot of grammar, spelling, and clarity issues.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+   
+Numba
+-----
+Numpy and Scipy provide rich but fixed set of mathematical instruments accelerated with C extensions. However sometimes, one might need non-standard math to be as fast as C extensions. That's where Numba :cite:`numba` can be efficiently used. Numba is a JIT compiler based on LLVM :cite:`llvm`. It aims to close the gap in performance between Python and statically typed, compiled languages like C/C ...
```
Spell out all acronyms (JIT) the first time used.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+-----
+Numpy and Scipy provide rich but fixed set of mathematical instruments accelerated with C extensions. However sometimes, one might need non-standard math to be as fast as C extensions. That's where Numba :cite:`numba` can be efficiently used. Numba is a JIT compiler based on LLVM :cite:`llvm`. It aims to close the gap in performance between Python and statically typed, compiled languages like C/C ...
+
+Numba implements the notion of ufunc defined in Scipy and extends it to a computation kernel which can be not only mapped onto arrays but also spread across multiple cores. The original Numba version implements it using pool of native threads and simple work-sharing scheduler which coordinates work distribution between them. If used in a parallel numeric Python application, it adds the third thre ...
```
What is a ufunc? You cite this earlier, but do not define it. User defined function?

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+.. code-block:: python
+    :linenos:
+
+    @nb.guvectorize('(f4[:],f4[:],f4[:],f4[:],f4[:]'\
```
Ooof this code is really terse and hard to understand.

antalmal replied:
it would probably distract readers from the main point (calling `numba.guvectorize` with `target='parallel'`). The code here is just for the best illustration of what the benchmark is doing. Do you think that  citation of Black Scholes formula or removing the code will help? @dalippa

dalippa replied:
If the main point is calling `numba.guvectorize` with `target='parallel'`, that should be highlighted in the text above the code block - and you should leave the code. Citing the Black Scholes formula would also be helpful on line 134 ("...runtimes perform with Black Scholes [_citation] implemented with...").

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+---------------------------
+Intel |R| TBB does not work well for blocking I/O operations because it limits number of active threads. It is applicable only for tasks which do not block in the operating system. Python module for Intel |R| TBB is in an experimental stage and might be not sufficiently optimized and verified with different use-cases. In particular, it does not yet use master thread efficiently as a regular TBB p ...
+
+Another limitation is that Intel |R| TBB coordinates threads only inside single process while the most popular approach to parallelism in Python is multi-processing. Though, Intel |R| TBB survives better than OpenMP in oversubscribed environment because it does not rely on particular the number of threads at any moment participating in parallel computation thus the threads preemted by the OS are  ...
```
"Though..." is awkward.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Another limitation is that Intel |R| TBB coordinates threads only inside single process while the most popular approach to parallelism in Python is multi-processing. Though, Intel |R| TBB survives better than OpenMP in oversubscribed environment because it does not rely on particular the number of threads at any moment participating in parallel computation thus the threads preemted by the OS are  ...
+
+On the other hand, slow adaption of Intel |R| TBB by Intel |R| MKL suggests to find and evaluate alternative ways such as implementation of restricted subset of OpenMP on top of TBB threads or vice-versa, OpenMP threads used as Intel |R| TBB workers. In both cases, we have prototypes with initial experemental data. Anoter approach is suggested by the observation that a moderate oversubscription,  ...
```
do you mean "adoption" instead of "adaption"?

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,177 @@
+
+Conclusion
+----------
+This paper described the issues of over-subscription and threading composability which affects performance of Python libraries and frameworks such as Numpy, Scipy, and Numba. Suggested solution is to use a common threading runtime such as Intel |R| TBB which limits number of threads in order to prevent oversubscription and coordinates parallel execution of independent program modules. Python modu ...
```
Should also mention bypassing the GIL.

usage in "Few examples show promising results where thanks to ..." is awkward and I don't think gets the intended point across. I think you want to get across the point that the examples you've seen so far show promising results. "Few examples show promising results" would indicate that your technique did *not* accomplish its goal.

antalmal replied:
Both bypassing the GIL and efficiency of multi-processing are not quite the focus of this paper.

I'm afraid I made too bold stress on its official status and what should be further fixed. This module is quite ready for the particular problems which it solves. And if we don't have a broader feedback and adoption, the whole set of issues will not get resolution while people keep using sub-optimal workarounds.

dalippa commented:
I think the core points of the paper are there, but the organization, grammar, spelling, and clarity are not where they need to be at the moment to recommend acceptance as of yet.

dalippa commented:
@antalmal Given the scope of some of the errors, I did not cite each specific grammar/spelling/clarity issue in comments. One thing that might make reviewing easier is to put each sentence on its own line in the text file, while having the compiled file look the way you want it to.

antalmal commented:
@dalippa, I agree that is not yet complete for publication. I also have few more ideas how to explain things there and the references are missed (I hope it is not that important at this time). And I've also arranged few internal reviews with native speakers and stakeholders.
Is there still some time to fix the paper for being accepted for Scipy proceedings? I do apologize that it takes that much time - I'm not professional writer as you can see :)

dalippa commented:
@antalmal I think that it *can* get ready in time; there is still 1 week before the final accept/reject deadline (Jul 10).

Usually, when submitting a paper to a conference, you get acceptance conditional on modifications (final deadline is Jul 30).

I realize you are not a native speaker, which is also why I was providing very detailed feedback, especially for spelling/grammatical errors, etc. I will also note that it is more typical to only receive more general feedback; just an FYI in case you submit work to conferences in the future.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+
+Scaling parallel programs is not an easy thing. There are two fundamental laws which mathematically describe and predict scalability of a program: Amdahl's Law and Gustafson-Barsis' Law [AmdVsGus]_.
+According to Amdahl's Law, speedup is limited by the serial portion of the work, which effectively puts a limit on scalability of parallel processing for a fixed-size job.
+Python is especially vulnerable to this because it makes the serial part of the same code much slower than if implemented in other languages due to its deeply dynamic and interpretative nature.
```
Since there are languages which are slower than python, I think rather than using `other languages` you should specify some language(s). 

Also, I feel replacing `than if implemented` with `compared to implementations` makes the sentence more clearer.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+Yet with another active thread pool in the application, it impairs better performance because while OpenMP worker threads keep consuming CPU time in busy-waiting loops, the other parallel work cannot start until OpenMP threads stop spinning or are preempted by the OS.
+
+Though overhead from linear over-subscription (e.g. 2x) is not always visible on the application level (especially for small systems), it can be tolerated in many cases when the work for parallel regions is big enough.
+However, the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region.
```
I think starting like this: `However, in the worst case a program ...` makes the sentence clearer.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+Though overhead from linear over-subscription (e.g. 2x) is not always visible on the application level (especially for small systems), it can be tolerated in many cases when the work for parallel regions is big enough.
+However, the worst case is when a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region.
+This results in quadratic over-subscription (with default settings) which ruins multi-threaded performance on systems with a significant number of threads (roughly more than ten).
+In some big systems, it may not even possible to create as many software threads as the number of hardware threads multiplied by itself; that many threads just eat up all the available resources.
```
it may not even possible -> it might not even be possible.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+
+.. class:: abstract
+
+   Python is popular among numeric communities that value it for easy to use number crunching modules like Numpy/Scipy, Dask, Numba, and many others.
```
Please add references for the packages.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+   
+   Execution times for QR validation example. :label:`qrpic`
+
+Figure :ref:`qrpic` shows times acquired on 32-core (no HT) machine with 64GB RAM.
```
Please write the full form of acronyms when using it for the first time.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+The reason why only 10 tasks were selected for this demonstration is as follows.
+If top-level parallelism can load all the available cores on the machine, there is not much sense using nested parallelism and Intel |R| TBB version shows no speedup over serial MKL version.
+In such cases, Intel |R| TBB could help by load-balancing at the end of the work, but this example is already quite balanced, so there is no visible difference.
+This example also models behavior of real-life applicationd which usually do not have enoght parallelism on the outermost level.
```
applicationd -> applications

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+
+Usage example
+-------------
+For our first experiment, we need Intel |R| Distribution for Python :cite:`intelpython` to be installed along with the Dask :cite:`dask` library which simplifies parallelism with Python.
```
Missing citations.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+Case study
+----------
+The previous example was intentionally selected to be small enough to fit into this paper with all the sources.
+Here is another case study :cite:`codefest` that is closer to real-world applications.
```
Missing citation.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+
+    Case study results: Generation of User Recommendations. :label:`casestudy`
+
+The leftmost result was acquired on pure, non-accelerated Python that comes by default on Fedora 23.
```
Maybe specify `leftmost result in Fig 3` ?

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+-----
+Numpy and Scipy provide a rich but fixed set of mathematical instruments accelerated with C extensions.
+However, sometimes one might need non-standard math to be as fast as C extensions.
+That's where Numba :cite:`numba` can be efficiently used.
```
Missing citations.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+Numpy and Scipy provide a rich but fixed set of mathematical instruments accelerated with C extensions.
+However, sometimes one might need non-standard math to be as fast as C extensions.
+That's where Numba :cite:`numba` can be efficiently used.
+Numba is a Just-In-Time compiler (JIT) based on LLVM :cite:`llvm`.
```
Missing citation.

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+Numba is a Just-In-Time compiler (JIT) based on LLVM :cite:`llvm`.
+It aims to close the gap in performance between Python and statically typed, compiled languages like C/C++, which also have popular implementation based on LLVM.
+
+Numba implements the notion of universal functions (ufunc, a scalar function which can be used for processing arrays as well) defined in Scipy :cite:`ufunc` and extends it to a computation kernel that can be not only mapped onto arrays but also spread the work across multiple cores.
```
`but also spread the -> but can also spread the` or `but also spread the work across -> but also spread across`

ankurankan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,314 @@
+
+    Black Scholes benchmark running with Numba on 32 threads. :label:`numbatbb`
+
+The Figure :ref:`numbatbb` shows how original Numba and TBB-based version perform with Black Scholes :cite:`bsform` benchmark implemented with Numba. The following code is a simplified version of this benchmark that gives an idea how to write parallel code using Numba:
```
Missing citation.

sarostru commented:
Due to the complication of the proceedings deadline lining up with everyone's travel schedule to attend the conference we've decided to allow authors to spend an additional 2 days, that is until July 12th to complete any major changes to the submission.  This may impact your enjoyment of the conference, so we hope not too many of you need to take the time.  For reviewers, we ask that you read over this final document before the conference is over and post your decision here in the comments on or before July 17th.

Thanks and have a great Scipy!
Your proceedings co-chairs Scott Rostrup (@sarostru) and Sebastian Benthall (@sbenthall)

antalmal commented:
The text is supposed to be final. Please remind me if I missed anything important from your previous comments, @kynan, @dalippa, @ankurankan. And thank you for your review work again!

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+.. class:: abstract
+
+   Python is popular among numeric communities that value it for easy to use number crunching modules like [NumPy]_, [SciPy]_, [Dask]_, [Numba]_, and many others.
+   These modules often use multi-threading for efficient parallelism (on a node) in order to utilize all the available CPU cores.
```
"on a node" is redundant.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+Motivation
+----------
+The fundamental shift toward parallelism was loudly declared more than 11 years ago [HSutter]_ and multi-core processors have become ubiquitous nowadays [WTichy]_.
+However, the software world changes slowly and Python along with its computing ecosystem is not an exception.
```
This is *not* the case. The software world actually changes very quickly; however, application developers have been slow to incorporate multi-threading more now that CPU speeds have flat-lined but the number of cores per CPU in mainstream chips has increased.

antalmal replied:
Ah, I see. You say that despite fast changes in software world generally it was slow to adopt multi-core parallelism, right? I disagree that it is limited to application developers if it implies that domain library/language developers are somewhat faster to adopt it. New program models are not easy to invent and not fast to adopt, especially because shared-memory parallelism adds complexity. Would adding `...changes slowly toward adoption of multi-core parallelism...` solve your concern? @dalippa

kynan replied:
... the adoption of multi-core parallelism in the software world has been slow ...

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+.. [WTichy]  Walter Tichy, "The Multicore Transformation", Ubiquity, Volume 2014 Issue May, May 2014. DOI: 10.1145/2618393.
+             http://ubiquity.acm.org/article.cfm?id=2618393
+
+The multi-processing type of parallelism is popular in Python but it is prone to inefficiency due to memory-related overhead.
```
Multi-processing includes both process-based and thread-based parallel computing. You may need to be more specific here "multi-processing using the built-in XYZ package is popular in Python..."

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+             http://ubiquity.acm.org/article.cfm?id=2618393
+
+The multi-processing type of parallelism is popular in Python but it is prone to inefficiency due to memory-related overhead.
+On the other hand, multi-threaded parallelism is known to be more efficient but with Python, it suffers from the limitations of the global interpreter lock [GIL]_, which prevents scaling of Python programs effectively serializing them.
```
run-on sentence: "it suffers from..." -> "though it suffers from"

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+On the other hand, multi-threaded parallelism is known to be more efficient but with Python, it suffers from the limitations of the global interpreter lock [GIL]_, which prevents scaling of Python programs effectively serializing them.
+However, when it comes to numeric computations, most of the time is spent in native code where the GIL can easily be released and programs can scale.
+
+.. [GIL] David Beazley, "Understanding the Python GIL", PyCON Python Conference, Atlanta, Georgia, 2010.
```
Are you using footnotes or endnotes for references?

antalmal replied:
These are end-notes. I checked with the generated paper.pdf

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+------------------
+Thus, the best strategy to efficiently load a multi-core system is still to fight against serial regions and synchronization.
+One way to do that is to expose parallelism on all the possible levels of an application.
+For example, make outermost loops parallel or explore functional or pipeline types of parallelism on the application level.
```
incomplete thought. No subject.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+For example, make outermost loops parallel or explore functional or pipeline types of parallelism on the application level.
+Python libraries that help to achieve this are Dask [Dask]_, Joblib [Joblib]_, and even the built-in :code:`multiprocessing` module [mproc]_ (including its :code:`ThreadPool` class).
+On the innermost level, data-parallelism can be delivered by Python modules like NumPy [Numpy]_ and SciPy [SciPy]_.
+These modules can be accelerated with an optimized math library like Intel |R| Math Kernel Library (Intel |R| MKL) [MKL]_, which is multi-threaded internally using OpenMP [OpenMP]_ by default.
```
"by default" is less necessary, unless you wish to state that it can be customized?

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+Issues of Over-subscription
+---------------------------
+Nevertheless, the libraries named above do not coordinate the creation or pooling of threads, which leads to the situation where there are more active software threads than available hardware resources.
```
Combine the two sentences: "... which leads to *over-subscription*, where there are more active..."

kynan replied:
which may lead

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+Because overhead from linear over-subscription (e.g. 2x) is not always visible on the application level (especially for small systems), it can be tolerated in many cases when the work for parallel regions is big enough.
+However, in the worst case a program starts multiple parallel tasks and each of these tasks ends up executing an OpenMP parallel region.
+This results in quadratic over-subscription (with default settings) which ruins multi-threaded performance on systems with a significant number of threads (roughly more than ten).
+In some big systems, it may not even be possible to create as many software threads as the number of hardware threads multiplied by itself; that many threads just eat up all the available resources.
```
remove "that many threads..." and change "... by itself" to "... by itself due to insufficient resources"

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+A perfectly composable component should be able to function efficiently among other such components without affecting their efficiency.
+The first aspect of building a composable threading system is to avoid creation of excessive software threads, preventing over-subscription.
+That effectively means that a component and especially a parallel region cannot dictate how many threads it needs for execution (*mandatory parallelism*).
+Instead, it should expose available parallelism to a work scheduler (*optional parallelism*), which is often implemented as a user-level work stealing task scheduler that coordinate tasks between components and parallel regions and map them onto software threads.
```
"that coordinates"

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+Intel's approach to achieve threading composability is to use Intel |R| Threading Building Blocks (Intel |R| TBB) library as the common work scheduler, see Figure :ref:`components`.
+Intel |R| TBB is an open-source, cross-platform, mature and recognized C++ library for enabling multi-core parallelism.
+It was designed for composability, as well as optional and nested parallelism support from its foundation.
```
"from its foundation" is unnecessary.

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+    # install Dask
+    conda install dask
+
+Now, let us write a simple program using NumPy that validates QR decomposition by multiplying computed components and comparing the result against the original input:
```
"Now, let us..." -> "The code below is a simple..."

dalippa commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+However, each Dask task executes the same NumPy matrix operations which are accelerated using Intel |R| MKL under the hood and thus multi-threaded by default.
+This combination results in nested parallelism, i.e. when one parallel component calls another component, which is also threaded.
+
+The reason why the Dask version was set to have only 10 tasks is to model real-life applications with limited parallelism on the outermost level, which is quite typical for functional and pipeline types of parallelism.
```
Isn't 10 a common number of threads (2n+2) since many processors now have 4 cores?

antalmal replied:
no. In such case, it results in over-subscription BTW. Though, for I/O bound apps, it is quite good approach.

dalippa commented:
@antalmal The paper is looking really good, and I recommend accepting the paper in the proceedings pending the following changes that can be made before the Jul 30 deadline:

* there are still narration (be consistent whether 1st person singular or plural), grammar, verbosity, and clarity issues
* wouldn't thrashing also be increased by over-subscription? I didn't see that mentioned.
* You mention "too many threads for the hardware resources available" but don't mention which resources. Do you mean all resources (memory, disk, network, etc) or just some?
* "limitations and future work" and "conclusion" can use some more polish. Conclusion should summarize the results (including actual values for increase in speed, etc.).

antalmal commented:
Cool, thank you @dalippa for recommending it! Is it final decision (for the current deadline) or @kynan should also decide to accept it?

If it is recommended for inclusion with the deadline for modifications extended to July 30th, I'll focus on my talk at the conference and house hunting in Austin now and then return to writing the paper and addressing your comments next week. Is it ok for you?

dalippa commented:
@antalmal The decision is pending the changes mentioned by me and any additional changes required by @kynan .

Carreau commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+Scaling parallel programs is not an easy thing. There are two fundamental laws which mathematically describe and predict scalability of a program: Amdahl's Law and Gustafson-Barsis' Law [AGlaws]_.
+According to Amdahl's Law, speedup is limited by the serial portion of the work, which effectively puts a limit on scalability of parallel processing for a fixed-size job.
+Python is especially vulnerable to this because it makes the serial part of the same code much slower compared to implementations  in typed languages due to its deeply dynamic and interpretative nature.
```
I'm usure about the  `compare to implementations in typed languages`, the statement after about the dynamicity is not false, the dynamicity prevent some optimisations due to the mutability of everything at all time. One wide known counter example is javascript, which is dynamic, weakly time, and is now super fast. 

The interpreted nature can be argued to be a detail of implementation of CPython. There a number of Jit implementation of Python.

See also Larry Hasting [gilectomy talk](https://www.youtube.com/watch?v=P3AyI_u66Bw) that show that the GIL is actually necesarry to make non parallel execution fast. 

I guess by "typed" language you mean "Ahead of time typing and compiling" which is often colloquially referred to as strong typing, and and though as mutually exclusive with dynamism. I think that can be rephrased more clearly.

antalmal replied:
Thanks, I'll clarify the typed languages claim.

I'm well aware of gilectomy, I have its t-shirt :) Do you want to suggest any modification w.r.t GIL?

Carreau commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+Intel |R| TBB is an open-source, cross-platform, mature and recognized C++ library for enabling multi-core parallelism.
+It was designed for composability, as well as optional and nested parallelism support from its foundation.
+
+In the Intel |R| Distribution for Python 2017 Beta and later, as part of Intel |R| TBB release 4.4 Update 5, I introduce an experimental module which unlocks the potential for additional performance for multi-threaded Python programs by enabling threading composability between two or more thread-enabled libraries.
```
`I introduce` is the only reference to you in person, were you meaning to use `Intel |R|` ?

Carreau replied:
Is `Intel |R| Distribution for Python 2017`  Python 2 or 3 , if both, which one are you benchmarking with ?

antalmal replied:
Ok, looks like nobody wants me to say it from 1st person, I'll fix that.
I did it with 3.5 for the paper. However, there is no difference in performance between Python 2 and 3 because cpython does not play significant role here. So, do you still want to add 'cpython v3.5.1' to the list of software versions?

Carreau replied:
> Ok, looks like nobody wants me to say it from 1st person

You can I have no issue with that, as it was the only one in the paper I though it might have been a mistake. 

> So, do you still want to add 'cpython v3.5.1' to the list of software versions?

Sure, at least it's explicit and people can reproduce. You can add a sentence saying that python versions has no effects.

Carreau commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+Usage example
+-------------
+For our first experiment, we need Intel |R| Distribution for Python [IntelPy]_ to be installed along with the Dask [Dask]_ library which simplifies parallelism with Python.
```
If you do an experiment can you pin/give the version(s) of the Dask library (and others) you use ?

antalmal replied:
I provided this info under each graph. Do you want it to be in the text as well?

Carreau commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+    assert(test.compute()) # threaded
+    print(time.time() - t0)
+
+Here, Dask splits the array into 10 chunks and processes them in parallel using multiple threads.
```
It's unclear what it's 10 here. Of course you divide you array in chuncks, but it would be nice to explain.

antalmal replied:
I did this below. If it is not enough, please elaborate.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+   Python is popular among numeric communities that value it for easy to use number crunching modules like [NumPy]_, [SciPy]_, [Dask]_, [Numba]_, and many others.
+   These modules often use multi-threading for efficient multi-core parallelism in order to utilize all the available CPU cores.
+   Nevertheless, their threads can interfere with each other leading to overhead and inefficiency if used together in one application.
+   The lost performance can still be recovered if all the multi-threaded parties are coordinated.
```
Loss of performance can be prevented

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+Gustafson-Barsis' law offers some hope stating that if the problem-size grows along with the number of parallel processors, while the serial portion grows slowly or remains fixed, speedup grows as processors are added.
+This might relax the concerns regarding Python as a language for parallel computing since the serial portion is mostly fixed in Python when all the data-processing is hidden behind libraries like NumPy and SciPy which are written in other languages.
+Nevertheless, a larger problem size demands more operational memory to be used for processing it, but memory is a limited resource.
+Even working with Big Data, it must be processed by chunks that fit into memory.
```
I don't think this sentence adds anything (only brings in the big data buzzword)

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+-----------------------
+Altogether, the co-existing issues of multi-threaded components define *threading composability* of a program module or a component.
+A perfectly composable component should be able to function efficiently among other such components without affecting their efficiency.
+The first aspect of building a composable threading system is to avoid creation of excessive software threads, preventing over-subscription.
```
of an excessive number of

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+   Execution times for QR validation example. :label:`qrpic`
+
+Figure :ref:`qrpic` shows performance results acquired on a 32-core (no hyper-threading) machine with 64GB memory.
+The Dask version runs slower than the NumPy version with the default setting because 10 outermost tasks end up calling 10 OpenMP-based parallel regions that create 10 times more threads than available hardware resources.
```
That is not quite true: 100 threads aren't quite 10x available resources on a 32 core machine.

antalmal replied:
I didn't say 100 threads. I said `10 times more than` 32, i.e. referring to 320 threads.

kynan replied:
You're right, I misread!

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+Here is the scalar function :code:`BlackScholes`, consisting of many elementary and transcendental operations, which is applied (*broadcasted*) by Numba to every element of the input arrays.
+Additionally, :code:`target='parallel'` specifies to run the computation using multiple threads.
+The real benchmark also computes the put price using :code:`numba.guvectorize`, uses approximated CND function instead of ERF for better SIMD optimization, optimizes the sequence of math operations for speed, and repeats the calculation multiple times.
```
Acronyms CND, ERF not explained

antalmal replied:
probably I'll rather remove mentioning of them completely. I assumed that these math functions are known.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+As was discussed above, the TBB-based implementation of Intel |R| MKL threading layer is yet in its infancy and is therefore suboptimal.
+However, all these problems can be eliminated as more users will become interested in solving their composability issues and Intel |R| MKL and the TBB module are further developed.
+Thus, please contact Intel in order to indicate your interest.
```
Such a request is rather odd in a paper...

dalippa replied:
I would agree. I think it might be reasonable to link to the download site for TBB (in a different place) and leave it at that.

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+
+A different approach is suggested by the observation that a moderate over-subscription, such as from two fully subscribed thread pools, does not significantly affect performance for most use cases.
+In this case, preventing quadratic over-subscription from the nested parallelism (in particular, with OpenMP) can be a practical alternative.
+Therefore, the solution for that can be as simple as "Global OpenMP Lock" (GOL) or a more elaborate inter-process semaphore that coordinates OpenMP parallel regions.
```
has this GOL been suggested before or did you come up with this?

antalmal replied:
I'm not aware of such suggestion before, I'll ask our OpenMP team if it was proposed before

kynan commented on papers/anton_malakhov/00_amalakhov.rst:
```diff
@@ -0,0 +1,335 @@
+This paper described the issues of multi-threaded programs and libraries such as GIL, over-subscription, and threading composability.
+These issues affect performance of Python libraries and frameworks such as NumPy, SciPy, and Numba.
+The suggested solution is to use a common threading runtime such as Intel |R| TBB which limits the number of threads in order to prevent over-subscription and coordinates parallel execution of independent program modules.
+Python module for Intel |R| TBB was implemented to substitute Python's ThreadPool implementation and switch Intel |R| MKL into TBB-based mode.
```
A Python module...

kynan commented:
I agree with @dalippa that this paper is now in a state where I can also recommend it for inclusion in the proceedings. Please address the remaining comments in the final version.

dalippa commented:
@antalmal Please let me and @kynan know when you're ready for the next re-review.

sarostru commented:
@dalippa @kynan Just so I understand, the current status is ready with some updates before the final version?  I'm adding the ready label, if that isn't right you can switch it to one of the other labels.

After the updates @antalmal, @sbenthall and I will ask for a final reread by yourself for proofreading and we will help with updating the build of the proceedings from draft to final.
Cheers,
Scott

dalippa commented:
@sarostru @sbenthall @antalmal My current understanding is that the paper is still pending a final review of the reviewers (to ensure comments are addressed), then by the proceedings chairs for final acceptance.

dalippa commented:
@antalmal Please let me know when you're ready for re-review.

antalmal commented:
@dalippa If it is urgent, I did the fixes requested already. But if I have few more time I hope to get back to conclusion section. Answering your remarks really quick now:

* 1st person statement was intentional though I'll look at this again.
* Theoretically thrashing is very likely with some cases with over-subscription. But I personally didn't experience that thus I'm not that certain to claim it. I can say "it might" probably.
* Re: resources availability, it depends on specific machine configuration what is the limiting factor why OpenMP cannot create that many threads. I might be max processes per user (`ulimit -u`) or it might be memory limitation because all these threads want big stacks to be allocated.

dalippa commented:
I don't think it is urgent but you would need to ask the conference chairs
about when they will be merging.

On Monday, August 1, 2016, Anton <notifications@github.com> wrote:

> @dalippa <https://github.com/dalippa> If it is urgent, I did the fixes
> requested already. But if I have few more time I hope to get back to
> conclusion section. Answering your remarks really quick now:
>
>    - 1st person statement was intentional though I'll look at this again.
>    - Theoretically thrashing is very likely with some cases with
>    over-subscription. But I personally didn't experience that thus I'm not
>    that certain to claim it. I can say "it might" probably.
>    - Re: resources availability, it depends on specific machine
>    configuration what is the limiting factor why OpenMP cannot create that
>    many threads. I might be max processes per user (ulimit -u) or it
>    might be memory limitation because all these threads want big stacks to be
>    allocated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/209#issuecomment-236773952>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHdFxcHagcUIyOEeCnoZg8pS27g88g3dks5qbqg7gaJpZM4IqOTG>
> .
>


-- 
Please forgive any typos since iSent this message from my iPad.

sarostru commented:
Likely not until the weekend, unless @sbenthall has time this week.  We would like to wrap things up as soon as we can though.

I'll remove the ready tag for now then, until you make those updates.

antalmal commented:
@dalippa I did several modifications. I hope these did not introduce new grammar mistakes..

dalippa commented:
@antalmal Re-reviewing now.

dalippa commented:
@antalmal I made a few comments, mostly regarding grammar.

antalmal commented:
@dalippa Thank you a lot! I also spotted incorrect percentage calculation for the 40% case. Fixed

sbenthall commented:
Reviewers: is this ready now?

Carreau commented:
> Reviewers: is this ready now?

It's ok with me. 

And side question – possibly for next year –  now that the talk itself [is online](https://www.youtube.com/watch?v=kfQcWez2URE), should we also link to it ? I guess that does not only apply for this paper.

stefanv commented:
All talks should be updated with a `:video:` link.  See, e.g.:

http://conference.scipy.org/proceedings/scipy2015/allen_downey.html

/cc @sbenthall

dalippa commented:
@antalmal Looks good now. When it's updated with the video, it can be merged.

@sarostru @sbenthall

sarostru commented:
Great, I'll merge it in and add the video link myself.
